File: assign13/q5.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 15 15:35:48 2023

@author: harsha
"""

class node:
    def __init__(self,data):
        self.data = data
        self.next = None
        
class ll:
    def __init__(self):
        self.head = None
        
def insert(root, item):
    temp = node(item)
      
    if (root == None):
        root = temp
    else :
        ptr = root
        while (ptr.next != None):
            ptr = ptr.next
        ptr.next = temp
      
    return root
        
def arraytoll(arr):
    root = None
    for i in range(len(arr)):
        root = insert(root, arr[i])
    return root

def display(root):
    while (root != None) :
        print(root.data, end = "->")
        root = root.next




def deletelast(head,key):
    cur = head;
    if(cur.data == key):prev= cur
    else:prev = None
    prev = None
    while(cur.next!=None):
        temp = cur.next
        if(temp.data == key and temp!=None):
            prev = cur
        cur = cur.next;
    prev.next = prev.next.next


    

link = ll()
link.head = node(1)
arr = [1,2,3,5,2,10]
root = arraytoll(arr)
deletelast(root,2)
display(root)

File: assign14/q5.cpp
#include <iostream>
#include <bits/stdc++.h>


using namespace std;

struct node{
	int data;
	node* next;
	//constrctor
	node(int x){
		this->data = x;
		this->next =NULL;
	}
};

node* array2ll(vector<int> &arr){
	node* head = new node(arr[0]);
	node* temp = head;
	for(int i=1;i<arr.size();i++){
		node* n = new node(arr[i]);
		temp->next = n;
		temp = temp->next;
	}
	return head;
}


void display(node *head){
	node *temp=head;
	while(temp!=NULL){
		cout<<temp->data<<"->";
		temp = temp->next;
	}
	cout<<endl;
}

//////////////  ANSWER /////////////////
node* oddandeven(node* head) {
    node* ans=head;
	node* temp = head->next;
    node* odd=head,*even = head->next;
    while(odd && odd->next && even && even->next){
        odd->next = odd->next->next;
        even->next = even->next->next;
		odd = odd->next;

		even = even->next;
    }
    odd->next = temp;
	return ans;
}

int main(){
	vector<int> v1{1,2,3,4,5,6,7,8};
	node* head = array2ll(v1);
	display(head); 
    node* ans = oddandeven(head);
	display(ans); 

	return 0;
}